python/python/lance/arrow.py

```diff
@@ -11,14 +11,182 @@
 #  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
-from typing import Optional, Union
+from typing import List, Optional, Union
 
 import pyarrow as pa
 
 from .lance import BFloat16
 from .lance import bfloat16_array as bfloat16_array
 
 
+class ImageURIType(pa.ExtensionType):
+    def __init__(self):
+        pa.ExtensionType.__init__(self, pa.string(), "lance.arrow.image_uri")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return ImageURIType()
+
+    def __arrow_ext_class__(self):
+        return ImageURIArray
+
+    def __arrow_ext_scalar_class__(self):
+        return ImageURIScalar
+
+
+class EncodedImageType(pa.ExtensionType):
+    def __init__(self):
+        pa.ExtensionType.__init__(self, pa.binary(), "lance.arrow.encoded_image")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return EncodedImageType()
+
+    def __arrow_ext_class__(self):
+        return EncodedImageArray
+
+    def __arrow_ext_scalar_class__(self):
+        return EncodedImageScalar
+
+
+class ImageTensorType(pa.ExtensionType):
+    def __init__(self, arrow_type, shape):
+        # TODO: use pa.VariableShapeTensorArray once available?
```

Comment:
Maybe we should not implement `ImageTensorType` at all until `VariableShapeTensorArray` is released. For now we could have encoded image type have a `to_numpy()` method to convert to a numpy tensor.

Reply:
On second thought, maybe we should rename to `FixedSizeImageTensorType` (what a mouthful...) and later add a variable shape.

As long as the user has a way to decode to a ragged numpy / tensorflow array themselves, I think we are good here.

Reply:
That's a good point - that would probably be best.
`VariableShapeTensorArray` will be available in arrow 14.0.0 at best (would require calling a vote this week and pushing the review).

Reply:
Changed to `FixedShapeImageTensorType`.